src/core/hash.rs

```diff
@@ -19,30 +23,123 @@ pub trait Hash: Debug + Clone + 'static {
         + PartialOrd;
 
     fn hash(s: &[u8]) -> Self::Output;
+
+    fn update(&mut self, s: &[u8]);
+
+    fn finalize(self) -> Self::Output;
+}
+
+#[derive(Debug, Clone)]
+pub struct Sha3Hasher(Option<CoreWrapper<Sha3_256Core>>);
+
+impl serde::ser::Serialize for Sha3Hasher {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str("sha3_256")
+    }
+}
+
+impl<'de> serde::de::Deserialize<'de> for Sha3Hasher {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct StringVisitor<T>(std::marker::PhantomData<fn() -> T>);
+
+        impl<'de, T> Visitor<'de> for StringVisitor<T>
+        where
+            T: std::str::FromStr,
+        {
+            type Value = T;
+
+            fn expecting(&self, formatter: &mut Formatter) -> std::fmt::Result {
+                write!(formatter, "sha3_256")
+            }
+
+            fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
+            where
+                E: Error,
+            {
+                FromStr::from_str(v).map_err(|_e| E::custom("hasher was badly mismatched"))
+            }
+        }
+
+        deserializer.deserialize_str(StringVisitor::<Sha3Hasher>(std::marker::PhantomData))
+    }
+}
+
+impl Default for Sha3Hasher {
+    fn default() -> Self {
+        Sha3Hasher(None)
+    }
+}
+
+impl Sha3Hasher {
+    pub fn new() -> Self {
+        Self(Some(Sha3_256::new()))
```

Comment:
Why do we use Option here?

Reply:
because `Hash` has two ways of use
```
let output  = Hash::hash(b"abc")
```
or 
```
let mut hash = Hash::new();
hash.update(b"a");
hash.update(b"b");
hash.update(b"c");
let output = hash.finalize();
```
So the `Option` maybe another way of expression which is equal with `MaybeUninit`.